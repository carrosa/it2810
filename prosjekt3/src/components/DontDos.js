import React from 'react';
import {View, AsyncStorage, ScrollView} from 'react-native';
import {
  Text,
  Card,
  Title,
  Paragraph,
  Checkbox,
  Caption,
  DarkTheme,
  DefaultTheme,
  Button,
  Divider
} from 'react-native-paper';
import Hr from 'react-native-hr-component';


const styles = {
  card: {
    flexDirection: 'row',
    marginLeft: 10,
    marginRight: 10,
    marginTop: 10,
    marginBottom: 10
  },
  caption: {
    marginTop: 20,
    marginBottom: 20,
    textAlign: 'center',
    fontSize: 20
  }
};

export default class DontDos extends React.Component {

  state = {
    tasks: {}
  };

  async componentDidMount() {
    /*
     * Henter inn antall To Don'ts
     * Henter inn alle Dont Dos og legger i state
     * Returnerer ingenting (exiter ut av funksjonen) hvis det ikke finnes noen toDonts (count < 1)
     */
    try {
      const tasks = await AsyncStorage.getItem('tasks');
      if (tasks) { // Sjekker om det finnes noen tasks, hvis ja så setter vi task state
        this.setState({tasks: JSON.parse(tasks)});
      }
    } catch (e) {
      console.error(e);
    }
  }

  render() {
    return (
        <ScrollView>
          <View>
            {
              this.createToDontCards()
            }
          </View>
        </ScrollView>
    );
  }

  async onCheckboxPress(e) {
    /*
     * Lag en ny state og sett nr i (den vi sendte inn) til motsatt verdien av det den var før
     * og så setstate med denne verdien. Dette for å oppdatere deler av state sånn at checkboxen blir checked
     */
    let newTaskState = this.state.tasks;
    newTaskState[e].done = !this.state.tasks[e].done;
    this.setState({tasks: newTaskState});
    await AsyncStorage.setItem('tasks', JSON.stringify(newTaskState));
  }

  // Hjelpemetode for å lage not completed to dont card
  createNotCompletedToDontCard(e, i) {
    return (
        <Card key={i} style={styles.card} elevation={5} theme={DefaultTheme}>
          <Card.Content>
            <Title>{this.state.tasks[e].title}</Title>
            <Paragraph>{this.state.tasks[e].content}</Paragraph>
          </Card.Content>
          <Divider/>
          <Card.Actions>
            <Button
                mode={'text'}
                onPress={() => this.onCheckboxPress(e)}
            >
              Marker som ferdig
            </Button>
          </Card.Actions>
        </Card>
    );
  }

  // Hjelpemetode for å lage completed toDontCard
  createCompletedToDontCard(e, i) {
    return (
        <Card key={'notDone' + i} style={styles.card} elevation={5} theme={DarkTheme}>
          <Card.Content>
            <Title>{this.state.tasks[e].title}</Title>
            <Paragraph>{this.state.tasks[e].content}</Paragraph>
          </Card.Content>
          <Divider/>
          <Card.Actions>
            <Button
                mode={'text'}
                onPress={() => this.onCheckboxPress(e)}
            >
              Marker som uferdig
            </Button>
          </Card.Actions>
        </Card>
    );
  }

  createToDontCards() {
    /*
     * Lager en komponent med info om det er en tom liste med tasks eller ikke
     */
    let empty = (isDoneString, key) => {
      return (
          [
            <Caption key={key} style={styles.caption}>
              Du har ingen {isDoneString} to donts
            </Caption>
          ]
      );
    };
    /*
     * Lager komponent av alle toDonts. Putter dette inn i render
     */
    let notDoneTasks = [<Hr key={'hrt'} lineColor='#000' thickness={3} fontSize={20} text='Ikke ferdig'/>];

    // Horisontal linje for å separere done og not done cards
    let doneTasks = [<Hr key={'hrb'} lineColor='#000' thickness={3} fontSize={20} text='Ferdig'/>];
    Object.keys(this.state.tasks).reverse().forEach((e, i) => { // Itererer baklengs for å få nyeste øverst
      /*
       * Sjekker om toDonten er null eller undefined just in case
       * (Det var noen som ble null da jeg lagde dette fordi jeg ikke parsa state til int når man laster
       * opp lastDontID med asyncstorage, men det er fiksa nå, så skal ikke være et problem det, men greit å ha
       * med en sjekk uansett tenker jeg. SKader ikke iallefall.
       */
      if (e !== null && e !== undefined) {
        if (!this.state.tasks[e].done) {
          notDoneTasks.push(this.createNotCompletedToDontCard(e, i));
        } else {
          doneTasks.push(this.createCompletedToDontCard(e, i));
        }
      }
    });
    return notDoneTasks.length > 1 ? notDoneTasks.concat(doneTasks.length > 1 ? doneTasks : doneTasks.concat(empty('fullførte', 'done')))
        : notDoneTasks.concat(empty('ufullførte', 'not done')).concat(doneTasks.length > 1 ? doneTasks : doneTasks.concat(empty('fullførte', 'done')));
  }
}