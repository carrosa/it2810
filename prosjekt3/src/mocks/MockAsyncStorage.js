/*
 * Klasse for å mocke AsyncStorage under testing
 * Alle funksjonene returnerer et promise sånn at vi kan teste async
 */
export default class MockAsyncStorage {
  constructor(cache = {}) {
    this.storageCache = cache;
  }

  setItem = jest.fn((key, value) => {
    return new Promise((resolve, reject) => {
      return (typeof key !== 'string' || typeof value !== 'string')
          ? reject(new Error('key and value must be string'))
          : resolve(this.storageCache[key] = value);
    });
  });

  getItem = jest.fn((key) => {
    return new Promise((resolve) => {
      return this.storageCache.hasOwnProperty(key)
          ? resolve(this.storageCache[key])
          : resolve(null);
    });
  });
}