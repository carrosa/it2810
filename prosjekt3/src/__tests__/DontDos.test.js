import React from 'react';
import renderer from 'react-test-renderer';
import DontDos from '../components/DontDos';
import MockAsyncStorage from '../mocks/MockAsyncStorage';


/*
 * Setter opp AsyncStorage Mock
 */
const storageCache = {};
const AsyncStorage = new MockAsyncStorage(storageCache);
jest.setMock('AsyncStorage', AsyncStorage);

// Task template for å putte inn i AsyncStorage
const task = (title, content, done) => {
  return {
    'title': title,
    'content': content,
    'done': done
  };
};


describe('<DontDos /> tests', () => {
  // Bygger DontDos
  const wrapper = renderer.create(<DontDos/>);
  const instance = wrapper.getInstance();

  // Test bygge snapshot
  test('<DontDos /> renders correctly', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  /*
   * Setter hva vi vil starte å ha i AsyncStorage og legger dette til i AsyncStorage
   * Dette for at vi skal ha Cards å teste
   * Gjør dette først, hvis ikke har ikke renderer noe å bygge
   */
  test('Add to AsyncStorage and test render', async () => {
    // Legger inn tasks i Async Storage
    const tasks = {};
    for (let i = 0; i < 5; i++) {
      tasks[i] = task(`Tittel ${i}`, `Content ${i}`, i % 2 === 0);
    }
    await AsyncStorage.setItem('tasks', JSON.stringify(tasks));
  });

  /*
   * Test createToDontCards()
   * Tester da også createNotCompletedToDontCard()
   * og createCompletedToDontCard()
   */
  test('Test: createToDontCards', async () => {
    // Må være async fordi componentDidMount er async
    // Må kalle componentDidMount for å sette state her, pga asyncstorage
    await instance.componentDidMount().then(() => {
      expect(instance.createToDontCards()).toMatchSnapshot();
    });
  });

  /*
   * Test onCheckboxPress(e)
   */
  test('Test: onCheckboxPress', async () => {
    // Endre fra done:true til done: false
    await instance.onCheckboxPress(0).then(async () => {
      // Sjekker at det er lagt i AsyncStorage
      expect(JSON.parse(await AsyncStorage.getItem('tasks'))[0].done).toBe(false);
      // Sjekker at state er forandret
      expect(instance.state.tasks[0].done).toBe(false);
    });
    // Endre fra done:false til done: true
    await instance.onCheckboxPress(1).then(async () => {
      // Sjekker at det er lagt i AsyncStorage
      expect(JSON.parse(await AsyncStorage.getItem('tasks'))[1].done).toBe(true);
      // Sjekker at state er forandret
      expect(instance.state.tasks[1].done).toBe(true);
    });
  });
});