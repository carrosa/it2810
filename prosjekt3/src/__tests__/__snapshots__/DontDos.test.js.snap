// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DontDos /> tests <DontDos /> renders correctly 1`] = `null`;

exports[`<DontDos /> tests Test: createToDontCards 1`] = `
Array [
  <Hr
    fontSize={20}
    lineColor="#000"
    text="Ikke ferdig"
    thickness={3}
  />,
  <withTheme(Card)
    elevation={5}
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#f6f6f6",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "error": "#B00020",
          "placeholder": "rgba(0, 0, 0, 0.54)",
          "primary": "#6200ee",
          "surface": "#ffffff",
          "text": "#000000",
        },
        "dark": false,
        "fonts": Object {
          "light": "HelveticaNeue-Light",
          "medium": "HelveticaNeue-Medium",
          "regular": "Helvetica Neue",
          "thin": "HelveticaNeue-Thin",
        },
        "roundness": 4,
      }
    }
  >
    <Card.Content>
      <Title>
        Tittel 3
      </Title>
      <Paragraph>
        Content 3
      </Paragraph>
    </Card.Content>
    <withTheme(Divider) />
    <Card.Actions>
      <withTheme(Button)
        mode="text"
        onPress={[Function]}
      >
        Marker som ferdig
      </withTheme(Button)>
    </Card.Actions>
  </withTheme(Card)>,
  <withTheme(Card)
    elevation={5}
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#f6f6f6",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "error": "#B00020",
          "placeholder": "rgba(0, 0, 0, 0.54)",
          "primary": "#6200ee",
          "surface": "#ffffff",
          "text": "#000000",
        },
        "dark": false,
        "fonts": Object {
          "light": "HelveticaNeue-Light",
          "medium": "HelveticaNeue-Medium",
          "regular": "Helvetica Neue",
          "thin": "HelveticaNeue-Thin",
        },
        "roundness": 4,
      }
    }
  >
    <Card.Content>
      <Title>
        Tittel 1
      </Title>
      <Paragraph>
        Content 1
      </Paragraph>
    </Card.Content>
    <withTheme(Divider) />
    <Card.Actions>
      <withTheme(Button)
        mode="text"
        onPress={[Function]}
      >
        Marker som ferdig
      </withTheme(Button)>
    </Card.Actions>
  </withTheme(Card)>,
  <Hr
    fontSize={20}
    lineColor="#000"
    text="Ferdig"
    thickness={3}
  />,
  <withTheme(Card)
    elevation={5}
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#242424",
          "disabled": "rgba(255, 255, 255, 0.3)",
          "error": "#ff1744",
          "placeholder": "rgba(255, 255, 255, 0.54)",
          "primary": "#03a9f4",
          "surface": "#424242",
          "text": "#ffffff",
        },
        "dark": true,
        "fonts": Object {
          "light": "HelveticaNeue-Light",
          "medium": "HelveticaNeue-Medium",
          "regular": "Helvetica Neue",
          "thin": "HelveticaNeue-Thin",
        },
        "roundness": 4,
      }
    }
  >
    <Card.Content>
      <Title>
        Tittel 4
      </Title>
      <Paragraph>
        Content 4
      </Paragraph>
    </Card.Content>
    <withTheme(Divider) />
    <Card.Actions>
      <withTheme(Button)
        mode="text"
        onPress={[Function]}
      >
        Marker som uferdig
      </withTheme(Button)>
    </Card.Actions>
  </withTheme(Card)>,
  <withTheme(Card)
    elevation={5}
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#242424",
          "disabled": "rgba(255, 255, 255, 0.3)",
          "error": "#ff1744",
          "placeholder": "rgba(255, 255, 255, 0.54)",
          "primary": "#03a9f4",
          "surface": "#424242",
          "text": "#ffffff",
        },
        "dark": true,
        "fonts": Object {
          "light": "HelveticaNeue-Light",
          "medium": "HelveticaNeue-Medium",
          "regular": "Helvetica Neue",
          "thin": "HelveticaNeue-Thin",
        },
        "roundness": 4,
      }
    }
  >
    <Card.Content>
      <Title>
        Tittel 2
      </Title>
      <Paragraph>
        Content 2
      </Paragraph>
    </Card.Content>
    <withTheme(Divider) />
    <Card.Actions>
      <withTheme(Button)
        mode="text"
        onPress={[Function]}
      >
        Marker som uferdig
      </withTheme(Button)>
    </Card.Actions>
  </withTheme(Card)>,
  <withTheme(Card)
    elevation={5}
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#242424",
          "disabled": "rgba(255, 255, 255, 0.3)",
          "error": "#ff1744",
          "placeholder": "rgba(255, 255, 255, 0.54)",
          "primary": "#03a9f4",
          "surface": "#424242",
          "text": "#ffffff",
        },
        "dark": true,
        "fonts": Object {
          "light": "HelveticaNeue-Light",
          "medium": "HelveticaNeue-Medium",
          "regular": "Helvetica Neue",
          "thin": "HelveticaNeue-Thin",
        },
        "roundness": 4,
      }
    }
  >
    <Card.Content>
      <Title>
        Tittel 0
      </Title>
      <Paragraph>
        Content 0
      </Paragraph>
    </Card.Content>
    <withTheme(Divider) />
    <Card.Actions>
      <withTheme(Button)
        mode="text"
        onPress={[Function]}
      >
        Marker som uferdig
      </withTheme(Button)>
    </Card.Actions>
  </withTheme(Card)>,
]
`;
